### PR TITLE
feat(strategy): add dynamic read write splitting

### LINKED ISSUE #194: feat(strategy): dynamic read write splitting
## Development Task
# Description
Now，Pisa-Proxy has supported static read write splitting strategy. Static strategy depends on config, once datasource status is changed, Pisa-Proxy can't route SQL correctly. In dynamic read write splitting strategy, Pisa-Proxy will probe datasource status and reconcile loadbalance strategy dynamicly.

# Implement
In this version, Pisa-proxy will support MHA high availability strategy. Pisa-Proxy will spawn four kind of monitor to probe the status of datasource.The rules match module will fork a thread to recive loadbalance strategy from dynamic read write splitting module by channel.In this module, there are four kind of monitor to probe datasource status.
- Connect Monitor: probe the connectivity  of datasource.
- Ping Monitor: probe health status of datasource.
- Lag Monitor: probe the late time between master node and slave node.
- ReadOnly Monitor: probe the role of datasource.

### Future Design Chart
image

#### Glossary
- Monitor Reconcile: Monitor Reconcile. Get status from monitors and compute final read write splitting strategy.
- Discovery: The kind of discovery, like MHA,RDS,MGR etc.
- Monitor: The kind of Monitor, includes Connect Monitor, Ping Monitor, ReadOnly Monitor, Lag Monitor.

### Probe Flow
1. Start Monitor to probe datasource.
2. Probe the connectivity of Master node and Slave node.
3. If connectivity is ok.
    3.1. probe the role of datasource.
    3.2. probe the late time between master node and slave node.
        3.2.1. If slave node is not late from master node, enter next time probe.
        3.2.2. If slave node is late from master node, update the list of loadbalance and enter next time probe.
4. If connectivity is not ok.
    4.1. If probe slave change to master,start Lag probe.
        4.1.1. If slave node is not late from master node, enter next time probe.
        4.1.2. If slave node is late from master node, update the list of loadbalance and enter next time probe.
    4.2.If slave node is not change to master, enter next time probe. 

image

### Configuration
|param | type| required| default | description|
|-- | -- | -- | -- | --|
|user | string | yes | None | Monitor user name|
|password | string | yes | None | Monitor password|
|monitor_interval | u64 | yes | 5000 | The interval of `Reconcile Monitor `update strategy(millisecond)|
|connect_interval | u64 | yes | 60000 | The interval of `Connect Monitor` probe (millisecond)|
|connect_timeout | u64 | yes| 600 | The timeout of `Connect Monitor(`millisecond)|
|connect_max_failures | u64 | yes | 3 | The max failures times of `Connect Monitor` probe|
|ping_interval | u64 | yes | 8000 | The interval of `Ping Monitor ` probe(millisecond)|
|ping_timeout | u64 | yes | 1000 | The timeout of `Ping Monitor` probe(millisecond)|
|ping_max_failures | u64 | yes | 3 | The max failures times of `Ping Monitor` probe|
|replication_lag_interval | u64 | yes | 10000 | The interval of `Lag Monitor` probe(millisecond)|
|replication_lag_timeout | u64 | yes | 1000| The timeout of `Lag Monitor` probe(millisecond)|
|replication_lag_max_failures | u64 | yes | 1 | The max failures of `Lag Monitor` probe|
|max_replication_lag | u64 | yes | 10000 |The threshold of `Lag Monitor` probe(millisecond)|
|read_only_interval | u64 | yes | 3000 | The interval of `ReadOnly Monitor` probe(millisecond)|
|read_only_timeout | u64 | yes | 600| The timeout of `ReadOnly Monitor` probe(millisecond)|
|read_only_max_failures | u64 | yes | 3 |The max failures times of `ReadOnly Monitor` probe|

##### Configuration Structure
```
#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct ReadWriteSplitting {
    #[serde(rename = "static")]
    pub statics: Option<ReadWriteSplittingStatic>,
    pub dynamic: Option<ReadWriteSplitting>
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ReadWriteSplittingDynamic {
    pub default_target: TargetRole,
    #[serde(rename = "rule")]
    pub rules: Vec<ReadWriteSplittingRule>,
    pub discovery: Discovery,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "lowercase", tag="type")]
pub enum Discovery {
    Mha(ManagerHighAvailability),
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
pub struct ManagerHighAvailability {
    pub user: String,
    pub password: String,
    pub pool_size: Option<u8>,
    pub monitor_interval: u64,
    pub connect_interval: u64,
    pub connect_timeout: u64,
    pub connect_max_failures: u64,
    pub ping_interval: u64,
    pub ping_timeout: u64,
    pub ping_max_failures: u64,
    pub replication_lag_interval: u64,
    pub replication_lag_timeout: u64,
    pub replication_lag_max_failures: u64,
    pub max_replication_lag: u64,
    pub read_only_interval: u64,
    pub read_only_timeout: u64,
    pub read_only_max_failures: u64,
}
```

##### Configuration Example
```
[proxy.config.read_write_splitting]

[proxy.config.read_write_splitting.dynamic]
default_target = "readwrite"

[proxy.config.read_write_splitting.dynamic.discovery]
type = "mha"
user = "monitor"
password = "monitor"
pool_size = 16
monitor_interval = 1000
connect_interval = 2000
connect_timeout = 200
connect_max_failures = 3
ping_interval = 1000
ping_timeout = 100
ping_max_failures = 3
replication_lag_interval = 1000
replication_lag_timeout = 3
replication_lag_max_failures = 3
max_replication_lag = 3
read_only_interval = 1000
read_only_timeout = 3
read_only_max_failures = 3

[[proxy.config.read_write_splitting.dynamic.rule]]
name = "write-rule"
type = "regex"
regex = ["^insert"]
target = "readwrite"
algorithm_name = "roundrobin"

[[proxy.config.read_write_splitting.dynamic.rule]]
name = "read-rule"
type = "regex"
regex = ["^select"]
target = "read"
algorithm_name = "roundrobin"
```

#### Check List
- [x] Monitor Reconcile.
- [x] configuration.
- [ ] Pisa-Controller.
- [x] Rules Match dynamic update.
- [x] Connect Monitor.
- [x] Ping Monitor.
- [x] Lag Monitor.
- [x] ReadOnly Monitor.
- [x] com query raw parse. @xuanyuan300 

Association Issue #88

### PR DESCRIPTION
**Pull request by wbtlb**
### What is changed and how it works?

Close #194 

What's Changed:
- add dynamic read write splitting

### CODE REVIEW

File: pisa-proxy/proxy/strategy/Cargo.toml [hunk 1]
```diff
@@ -7,5 +7,14 @@ edition = "2021"
...
+chrono = "0.4"
+tokio = { version = "1.0.0", features = ["full", "tracing"] }
+tokio-scoped = "0.2.0"
 loadbalance = { path = "../loadbalance" }
-endpoint = { path = "../endpoint" }
\ No newline at end of file
+endpoint = { path = "../endpoint" }
+conn_pool = { path = "../../proxy/pool" }
+mysql_protocol = {path = "../../protocol/mysql", package = "mysql_protocol"}
+pisa_error = {path = "../../error", package = "error"}
```
Comment (dongzl) [inline pisa-proxy/proxy/strategy/Cargo.toml:20]: suggest add blank line at end of file.

File: pisa-proxy/runtime/mysql/Cargo.toml [hunk 2]
```diff
@@ -30,5 +30,4 @@ endpoint = {path = "../../proxy/endpoint"}
 mysql_parser = {path = "../../parser/mysql"}
 common = {path = "../common"}
 plugin = {path = "../../plugin" }
-strategy = {path = "../../proxy/strategy"}
-
+strategy = {path = "../../proxy/strategy"}
```
Comment (dongzl) [inline pisa-proxy/runtime/mysql/Cargo.toml:33]: suggest add blank line at end of file.

File: pisa-proxy/Cargo.lock.bak [hunk 3]
```diff
@@ -0,0 +1,4065 @@
+# This file is automatically @generated by Cargo.
```
Comment (xuanyuan300) [inline pisa-proxy/Cargo.lock.bak:1]: remove this file

File: pisa-proxy/protocol/mysql/src/value.rs [hunk 4]
```diff
@@ -48,10 +55,21 @@ impl Convert<String> for String {
 impl Convert<u64> for u64 {
     fn new(val: Vec<u8>) -> Result<u64> {
         let mut buf = val.as_slice();
+        // println!("u64 >> {:#?}", val.get_u64_be());
```
Comment (xuanyuan300) [inline pisa-proxy/protocol/mysql/src/value.rs:58]: remove this comments

File: pisa-proxy/protocol/mysql/src/value.rs [hunk 5]
```diff
@@ -48,10 +55,21 @@ impl Convert<String> for String {
...
     }
 }
 
+impl Convert<Option<u64>> for Option<u64> {
+    fn new(val: Vec<u8>) -> Result<Option<u64>> {
+        let buf = val.as_slice();
+        if buf[0] == 0 {
+            return Ok(None);
+        }
+        Ok(Some(String::from_utf8(val).unwrap().parse::<u64>().unwrap()))
```
Comment (xuanyuan300) [inline pisa-proxy/protocol/mysql/src/value.rs:69]: all `unwrap` should be changed `?` .

File: pisa-proxy/proxy/strategy/src/readwritesplitting/rule_match.rs [hunk 6]
```diff
@@ -77,14 +83,39 @@ impl RulesMatchBuilder {
...
+    pub async fn start_rules_match_reconcile(
+        rx: crossbeam_channel::Receiver<ReadWriteEndpoint>,
+        inner: Arc<parking_lot::Mutex<RulesMatch>>,
+        rules: Vec<ReadWriteSplittingRule>,
+        default_target: TargetRole,
+    ) {
+        tokio::spawn(async move {
+            loop {
+                let rw_endpoint = rx.clone().recv().unwrap();
+                inner.clone().lock().default_target = default_target.clone();
```
Comment (xuanyuan300) [inline pisa-proxy/proxy/strategy/src/readwritesplitting/rule_match.rs:103]: Reassign default_target whether may be removed here?

File: pisa-proxy/proxy/strategy/Cargo.toml (hunk 1 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/Cargo.toml:24]: fixed

File: pisa-proxy/runtime/mysql/Cargo.toml (hunk 2 above)
Comment (wbtlb) [reply pisa-proxy/runtime/mysql/Cargo.toml:33]: fixed

File: pisa-proxy/protocol/mysql/src/client/conn.rs [hunk 7]
```diff
@@ -192,6 +192,28 @@ impl ClientConn {
...
     }
 
+    pub async fn send_ping(&mut self) -> Result<bool, ProtocolError> {
+        let framed = self.framed.take().unwrap();
+        let mut common_codec = framed.into_common();
+
+        common_codec.send((COM_PING, &[])).await?;
+        let res = match common_codec.next().await {
+            Some(Ok(data)) => {
+                if data.0[4] == OK_HEADER {
```
Comment (mlycore) [inline pisa-proxy/protocol/mysql/src/client/conn.rs:202]: Make it a general function if it is a common need.

File: pisa-proxy/protocol/mysql/src/value.rs [hunk 8]
```diff
@@ -52,6 +59,19 @@ impl Convert<u64> for u64 {
...
+impl Convert<Option<u64>> for Option<u64> {
+    fn new(val: Vec<u8>) -> Result<Option<u64>> {
+        let buf = val.as_slice();
+        if buf[0] == 0 {
+            return Ok(None);
+        }
+        let v = String::from_utf8(val)?;
+        let result = v.parse::<u64>()?;
+        Ok(Some(result))
+        // Ok(Some(String::from_utf8(val).unwrap().parse::<u64>().unwrap()))
```
Comment (mlycore) [inline pisa-proxy/protocol/mysql/src/value.rs:71]: Remove this comment

File: pisa-proxy/proxy/strategy/src/discovery/discovery.rs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    fn build(config: MasterHighAvailability, rw_endpoint: ReadWriteEndpoint) -> Self::Output {
+        Self { config, rw_endpoint, monitors: vec![] }
+    }
+
+    fn build_monitors(
+        &self,
+        monitor_response_channel: crate::readwritesplitting::MonitorResponseChannel,
+    ) -> Vec<MonitorKind> {
+        let mut monitors = vec![];
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/discovery/discovery.rs:56]: Could make this a factory in next release.

File: pisa-proxy/proxy/strategy/src/discovery/discovery.rs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+        monitors
+    }
+}
+
+#[derive(Debug)]
+pub enum MonitorKind {
+    Connect(MonitorConnect),
+    Ping(MonitorPing),
+    Lag(MonitorReplicationLag),
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/discovery/discovery.rs:103]: Better keep it same with name ReplicationLag

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    // probe datasource by connect
+    pub async fn connnect_check(endpoint: String) -> ConnectStatus {
+        let factory =
+            ClientConn::with_opts(String::from("root"), String::from("12345678"), endpoint);
+        match factory.connect().await {
+            Ok(_) => return ConnectStatus::Connected,
+            Err(_) => return ConnectStatus::Disconnected,
+        }
+
+        // match tokio::net::TcpStream::connect(endpoint.clone()).await {
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs]: Remove this comments

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+            connect_timeout,
+            connect_failure_threshold,
+            rw_endpoint,
+        }
+    }
+
+    // probe datasource by connect
+    pub async fn connnect_check(endpoint: String) -> ConnectStatus {
+        let factory =
+            ClientConn::with_opts(String::from("root"), String::from("12345678"), endpoint);
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs]: Is this hard-coded test data ?

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+#[async_trait::async_trait]
+impl Monitor for MonitorConnect {
+    async fn run_check(&self) {
+        let connect_period = self.connect_period;
+        let connect_failure_threshold = self.connect_failure_threshold;
+        let connect_timeout = self.connect_timeout;
+        let rw_endpoint = self.rw_endpoint.clone();
+        // let connect_tx = self.connect_tx.clone();
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs:106]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+};
+
+#[derive(Debug)]
+pub struct MonitorPing {
+    pub user: String,
+    pub password: String,
+    pub ping_period: u64,
+    pub ping_timeout: u64,
+    pub ping_failure_threshold: u64,
+    // pub ping_tx: crossbeam_channel::Sender<PingMonitorResponse>,
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs:34]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+impl MonitorPing {
+    pub fn new(
+        user: String,
+        password: String,
+        ping_period: u64,
+        ping_timeout: u64,
+        ping_failure_threshold: u64,
+        // ping_tx: crossbeam_channel::Sender<PingMonitorResponse>,
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs:72]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+                        debug!("ping monitor check timeout");
+                    }
+                }
+
+                if let Err(err) =
+                    monitor_response_tx.send(MonitorResponse::PingMonitorResponse(response.clone()))
+                {
+                    error!("send ping response err: {:#?}", err.into_inner());
+                }
+                // if let Err(e) = ping_tx.send(response.clone()) {
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+};
+
+#[derive(Debug)]
+pub struct MonitorReadOnly {
+    pub user: String,
+    pub password: String,
+    pub read_only_period: u64,
+    pub read_only_timeout: u64,
+    pub read_only_failure_threshold: u64,
+    // pub read_only_tx: crossbeam_channel::Sender<ReadOnlyMonitorResponse>,
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs:35]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+impl MonitorReadOnly {
+    pub fn new(
+        user: String,
+        password: String,
+        read_only_period: u64,
+        read_only_timeout: u64,
+        read_only_failure_threshold: u64,
+        // read_only_tx: crossbeam_channel::Sender<ReadOnlyMonitorResponse>,
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs:72]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+                    {
+                        debug!("read only monitor check timeout");
+                    };
+                }
+                if let Err(err) = monitor_response_tx
+                    .send(MonitorResponse::ReadOnlyMonitorResponse(response.clone()))
+                {
+                    error!("send read only response err: {:#?}", err.into_inner());
+                }
+                // if let Err(e) = read_only_tx.send(response.clone()) {
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+
+impl MonitorReplicationLag {
+    pub fn new(
+        user: String,
+        password: String,
+        replication_lag_period: u64,
+        replication_lag_timeout: u64,
+        replication_lag_failure_threshold: u64,
+        max_replication_lag: u64,
+        // replication_lag_tx: crossbeam_channel::Sender<ReplicationLagMonitorResponse>,
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:50]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+        let mut response = ReplicationLagMonitorResponse::new(curr_rw_endpoint.clone());
+
+        tokio::spawn(async move {
+            let mut retries = 1;
+            loop {
+                if curr_rw_endpoint.read.len() == 0 {
+                    monitor_response_tx
+                        .send(MonitorResponse::ReplicationLagResponse(response.clone()))
+                        .unwrap();
+                    // replication_lag_tx.send(response.clone()).unwrap();
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:191]: Remove unused comment codes

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+                                    user.clone(),
+                                    password.clone(),
+                                    read.addr.clone(),
+                                )
+                                .await
+                                {
+                                    Ok(lag) => {
+                                        match lag {
+                                            Some(lag) => {
+                                                // if slave lag > customer lag threshold
```
Comment (mlycore) [inline pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:209]: Remove unused comment codes

File: pisa-proxy/runtime/mysql/src/server/server.rs [hunk 23]
```diff
@@ -417,7 +417,6 @@ impl MySQLServer {
         //    error!("err:{:?}", err);
         //}
         //let mut client_conn = self.trans_fsm.get_conn().await.unwrap();
```
Comment (mlycore) [inline pisa-proxy/runtime/mysql/src/server/server.rs:419]: Remove unused comment codes

File: pisa-proxy/protocol/mysql/src/client/conn.rs (hunk 7 above)
Comment (wbtlb) [reply pisa-proxy/protocol/mysql/src/client/conn.rs:202]: fix with adding `is_ok_header() `method

File: pisa-proxy/protocol/mysql/src/value.rs (hunk 8 above)
Comment (wbtlb) [reply pisa-proxy/protocol/mysql/src/value.rs:71]: removed

File: pisa-proxy/proxy/strategy/src/discovery/discovery.rs (hunk 9 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/discovery/discovery.rs:56]: mark

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs (hunk 11 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs]: removed

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs (hunk 12 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs]: fixed

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs (hunk 13 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs:106]: removed

File: pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+                        debug!("connect monitor check timeout");
+                    };
+                }
+
+                if let Err(err) = monitor_response_tx
+                    .send(MonitorResponse::ConnectMonitorResponse(response.clone()))
+                {
+                    error!("send connect response err: {:#?}", err.into_inner());
+                }
+                // if let Err(e) = connect_tx.send(response.clone()) {
```
[the remark this answers is not in the document]
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/connect_monitor.rs]: removed

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs (hunk 14 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs:34]: removed

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs (hunk 15 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs:72]: removed

File: pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs (hunk 16 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/ping_monitor.rs]: removed

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs (hunk 17 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs:35]: removed

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs (hunk 18 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs:72]: removed

File: pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs (hunk 19 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/read_only_monitor.rs]: removed

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs (hunk 20 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:50]: removed

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs (hunk 21 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:191]: removed

File: pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs (hunk 22 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/monitors/replication_lag_monitor.rs:209]: removed

File: pisa-proxy/runtime/mysql/src/server/server.rs (hunk 23 above)
Comment (wbtlb) [reply pisa-proxy/runtime/mysql/src/server/server.rs:419]: removed

File: pisa-proxy/proxy/strategy/src/discovery/discovery.rs (hunk 10 above)
Comment (wbtlb) [reply pisa-proxy/proxy/strategy/src/discovery/discovery.rs:103]: update to ReplicationLag

**Review by mlycore — APPROVED** [review]

**Pull request merged**